### ISSUE TITLE
Multilabel Option for Accuracy, Precision and Recall

### ISSUE THREAD

**Issue by anmolsjoshi** [pytorch/ignite#310]
### Proposal
As previously discussed #275 , we should incorporate multilabel cases for precision, recall and accuracy. 

### How
Add a multilabel flag to each of the metrics, listed above. For multilabel cases, y and y_pred should be shape (batch_size, n_classes, ...) where n_classes is more than 1 (not possible with binary). 

Convert y_pred to 0's and 1's using the method below or round y_pred.

```
y_pred = y_pred.unsqueeze(1)
y_pred = torch.cat([1.0 - y_pred, y_pred], dim=1)
y_pred = torch.max(y_pred, dim=1)[1]
```

Convert y_pred and y to a shape of (-1, n_classes), to handle classification, NER, etc. 

For accuracy, use torch.eq to compare every item of y and y_pred. Count number of True. 

For precision and recall, keep the calculations the same i.e. calculating y * y_pred

Thoughts?

**Comment by anmolsjoshi** [comment]
@vfdev-5 thoughts?

**Comment by vfdev-5** [comment]
> Convert y_pred to 0's and 1's using the method below or round y_pred.

Maybe this should be configurable with a function ?

> Convert y_pred and y to a shape of (-1, n_classes), to handle classification, NER, etc.

👍 

```
y_pred = y_pred.unsqueeze(1)
y_pred = torch.cat([1.0 - y_pred, y_pred], dim=1)
y_pred = torch.max(y_pred, dim=1)[1]
```
This I do not quite understand why.


Thinking about implementation, I think we would need to define another update function and not try to 
incorporate a lot of `if self._is_multilabel`.
Maybe we can create a class `MultilabelAccuracy` in `accuracy.py` near `Accuracy` and maybe inherit of the latter.

**Comment by anmolsjoshi** [comment]
@vfdev-5 the snippet of code is another method to convert y_pred to 1's and 0's and return the same shape as y. please feel free to ignore it, we can stick with torch.round as the default function and allow it to be overridden by the user (different threshold, etc).

> Maybe we can create a class MultilabelAccuracy in accuracy.py near Accuracy and maybe inherit of the latter

Did you mean MultilabelAccuracy would be another metric? available in accuracy.py and inheriting Accuracy, with a changed update function. so a user would call MultilabelAccuracy directly? not using a is_multilabel option.

**Comment by vfdev-5** [comment]
> Did you mean MultilabelAccuracy would be another metric? available in accuracy.py and inheriting Accuracy, with a changed update function. so a user would call MultilabelAccuracy directly? not using a is_multilabel option.

@anmolsjoshi correct. I thought about something like: `from ignite.metrics.accuracy import Accuracy, MultilabelAccuracy`. I was trying to implement update function and it becomes a mess of `if else`. If you can find a more neat solution, we'll take it.

Just to make sure, `y_pred` is of shape (n_batch, n_classes, ...) and is composed of logit or probas etc and a priori is not bound between 0 and 1. However, `y_true` is of shape (n_batch, n_classes, ...) of type `long` and should satisfy `y_true ** 2 == y_true` <=> is composed of 0's and 1's.

First step we need to transform `y_pred` to 0's and 1's (using user provided rule or round), keeping shape (n_batch, n_classes, ...).
Second we compute number of whatever needed depending on if it is accuracy, recall or precision

**Comment by vfdev-5** [comment]
cc @rpatrik96

**Comment by vfdev-5** [comment]
@anmolsjoshi I think it is different for multilabel precision and recall too, but we need to verify. Take a look at [sklearn.precision](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_score.html#sklearn.metrics.precision_score) as reference.
To compute with multilabels it will be something like : `precision_score(y_true, np.round(y_pred), average='samples')`.

**Comment by anmolsjoshi** [comment]
@vfdev-5 in that example you wrote, if we calculate np.round(y_pred) * y_true and take the sum we get 22 - that is the total number of 1's in the intersection of y_pred and y.

np.round(y_pred).sum() gives us 30 - that's the total number of 1's in y_pred. 

Dividing these two gives us 0.7333. But when tested with sklearn, the answer is 0.7435. 

I will have to investigate this further.

**Comment by anmolsjoshi** [comment]
@vfdev-5 thanks for your response! I went through the [sklearn docs](https://scikit-learn.org/dev/modules/generated/sklearn.metrics.multilabel_confusion_matrix.html) and [code](https://github.com/scikit-learn/scikit-learn/blob/02dc9ed/sklearn/metrics/classification.py#L298) for their dev branch and have a path to move forward this. 

It seems that for multilabel cases, this is how sklearn handles it - 
- a confusion matrix is calculated for each sample in the batch
- precision/recall are calculated for each confusion matrix
- these metrics are averaged over each sample. 

It's pretty much how you said it. 

I'll work on this tonight and hopefully have something concrete.
